### PR TITLE
Automatically generate an issue for each `C-Needs-Release-Note` PR merged to Bevy

### LINKED ISSUE #1220: Automatically generate issues for each uncompleted release note
Two of the largest remaining challenges involved in #1163 are:

1. Prompting feature authors to write release notes for the corresponding sections.
2. Keeping track of what release note sections still need to be written, especially as more sections are completed.

[Previously](https://github.com/bevyengine/bevy-website/pull/891#issuecomment-1914655314), this was done by manual suggestions and tracking. This is a) a lot of work to generate and b) generates an endless stream of notifications for everyone involved.

There are two proposals for how we could handle and automate this.

# Proposal 1: big tracking issue

As part of a tracking issue for the release (e.g. #1188), we could generate a checkbox list like before, suggesting that all of the authors listed on the PR handle the process.

## Advantages

- no new issues generated.

## Disadvantages

- everyone gets pinged for every update
- the thread gets very long as people chime in to claim issues
- only 50 people can be pinged in a single comment
- only maintainers and the comment author can update the status of each ticket
- discoverability is poor
- doesn't work terribly well when done on a continual basis: we need to make the tracking issue for the next release immediately and constantly edit it?

# Proposal 2: a million issues

Instead, for each release note entry generated (aka each merged PR with a `C-Needs-Release-Note` entry, a bot makes an issue with a  "Write release notes for PR TITLE" with a link to the PR and pings the authors. This issue is added to the milestone.

## Advantages

- discussions are separated, and thus much less noisy
- much easier to discover work that needs to be done
- PRs can say "Closes #X" or the triage team can update this and it will work great without maintainer intervention
- we can add advice on how to write release notes to every issue
- issues can be generated dynamically as PRs are merged throughout the cycle

## Disadvantages

- makes a large number of issues! This will be noisy for 0.14 due to the last-minute nature of the work
- need to figure out more complex bot workflows to write this tool

### PR DESCRIPTION
**Pull request by alice-i-cecile**
Fixes #1220.

As discussed there, this PR will cause running the `generate-release` tool to automatically generate an issue for each release note section, pinging the authors to help write it.

While this will be noisy initially, this should be very helpful in coordinating the release process, and ease the notification burden on individual contributors by providing separate places where they can discuss each release note.

## Sample output

See https://github.com/alice-i-cecile/bevy-website/issues/10

## To do

- [x] make sure not to open duplicate issues
- [x] get the full list of issues, not just one
- [x] fix JSON conversion error
- [x] validate that duplication checking works as intended
- [x] validate that issues can be created successfully

## Future work

- look up the correct milestone for the release, and add the generated issue to it
- add a daily job that runs this workflow, opening PRs to update the release notes if needed

### CODE REVIEW

**Comment by BD103** [comment]
Managed to debug while this was erroring: It was due to some issues having `"closed_at": null`. We only expected a `DateTime`, not an `Option<DateTime>`. I fixed it with this patch, though I'm not sure if the `.unwrap()` logic I used was correct:

```patch
diff --git a/generate-release/src/github_client.rs b/generate-release/src/github_client.rs
index 3ad666c..c898803 100644
--- a/generate-release/src/github_client.rs
+++ b/generate-release/src/github_client.rs
@@ -96,7 +96,7 @@ pub struct GithubIssuesResponse {
     pub body: Option<String>,
     pub labels: Vec<GithubLabel>,
     pub user: GithubUser,
-    pub closed_at: DateTime<Utc>,
+    pub closed_at: Option<DateTime<Utc>>,
     pub pull_request: Option<GithubIssuesResponsePullRequest>,
 }
 
@@ -204,11 +204,16 @@ impl GithubClient {
             page += 1;
             if let Some(pr) = prs.last() {
                 if let Some(datetime_utc) = datetime_utc {
-                    if pr.closed_at < datetime_utc {
-                        println!(
-                            "\x1b[93mSkipping PR closed before the target datetime {}\x1b[0m",
-                            pr.closed_at
-                        );
+                    if let Some(closed_at) = pr.closed_at {
+                        if closed_at < datetime_utc {
+                            println!(
+                                "\x1b[93mSkipping PR closed before the target datetime {}\x1b[0m",
+                                closed_at
+                            );
+                            break;
+                        }
+                    } else {
+                        println!("Skipping open PR.");
                         break;
                     }
                 }
@@ -222,7 +227,7 @@ impl GithubClient {
                 datetime_utc
             );
 
-            prs.retain(|pr| pr.closed_at < datetime_utc);
+            prs.retain(|pr| pr.closed_at.is_some_and(|closed_at| closed_at < datetime_utc));
         }
 
         Ok(prs)

```

File: generate-release/src/release_notes.rs
Comment (BD103) [inline generate-release/src/release_notes.rs:182]:
```suggestion
        Please reply below if you'd like to volunteer do so, even if you're the author of the PR.
```

This version is a little less wordy, while still clear that anyone can contribute.

File: generate-release/src/release_notes.rs
Comment (BD103) [inline generate-release/src/release_notes.rs:194]:
```suggestion
        To submit your release notes, modify `{file_path}` and submit a PR.
```

`.md` is already in this filename, so it gets duplicated with this!

File: generate-release/src/release_notes.rs
Comment (BD103) [inline generate-release/src/release_notes.rs:178]:
```suggestion
    let authors = authors.join(", ");
```

Clippy doesn't like this line, so here's a simplified version.

**Comment by alice-i-cecile** [comment]
Issues can't be created correctly currently: we need to double check the API docs and make sure we're hitting the right end point.

File: release-content/0.14/release-notes/_release-notes.toml
Comment (alice-i-cecile) [inline release-content/0.14/release-notes/_release-notes.toml:1]: This should not have been changed.

File: generate-release/src/release_notes.rs
Comment (MrGVSV) [inline generate-release/src/release_notes.rs:193]: Nit: we should probably indicate why we're pinging these users. Maybe something like:
```suggestion
        Original Authors: {authors}
```

Comment (MrGVSV) [inline generate-release/src/release_notes.rs:193]: I wonder if we should give some indication here that the PR authors are welcome to write the release notes themselves, act as a consultant, or do nothing if they don't want to participate in this step for whatever reason.

I feel like it's sorta implied already, but should we make it more explicit?

**Review by jgayfer — APPROVED** [review]
I haven't tested this, but went through it all and looks sane to me.

File: generate-release/src/main.rs
Comment (BD103) [inline generate-release/src/main.rs:99]: (Nit) Could you add docs here describing that this command creates a bunch of issues? (Also recommend that users add `--dry-run` when testing, so we don't get spammed.)

**Review by BD103 — APPROVED** [review]
Looks good! (Would it be possible to test creating only a single issue, to ensure the output looks good in Github? If not then that's fine.)

**Comment by alice-i-cecile** [comment]
https://github.com/bevyengine/bevy-website/issues/1225 :fire:

**Comment by alice-i-cecile** [comment]
Duplication checking isn't working properly: see #1267 and #1247.

File: generate-release/src/release_notes.rs
Comment (MrGVSV) [inline generate-release/src/release_notes.rs:173]: I'm pretty sure GitHub-flavored Markdown doesn't ignore newlines, so this is actually spread across two separate lines.
```suggestion
While the author(s) of the PR often have the context, knowledge and motivation to draft the release notes for their feature, anyone can contribute release notes!
```

Comment (MrGVSV) [reply generate-release/src/release_notes.rs:173]: Yeah it does separate the lines: 
Screenshot of the text being rendered across two lines instead of one

File: generate-release/src/github_client.rs
Comment (alice-i-cecile) [reply generate-release/src/github_client.rs:367]:
```suggestion
                // TODO: add the milestone. Tracked in https://github.com/bevyengine/bevy-website/issues/1269
```

File: generate-release/src/release_notes.rs
Comment (alice-i-cecile) [reply generate-release/src/release_notes.rs:157]: I'm not concerned enough about that happening a) within the cycle and b) after the PR is already merged to try and do a more complex deduplication strategy. At least in this PR :)

We could parse it to grab the number I suppose?

Comment (MrGVSV) [reply generate-release/src/release_notes.rs:157]: Yeah definitely not a blocker for this PR but possibly something to consider for future work

**Review by MrGVSV — APPROVED** [review]

**Review by mockersf — CHANGES_REQUESTED** [review]
we can't automatically mention people, that is a very bad idea and most likely against GitHub terms of services
automated mentions without having explicit consent if spam

File: generate-release/src/github_client.rs
Comment (BD103) [inline generate-release/src/github_client.rs:402]:
```suggestion
#[derive(Debug, Clone, Copy)]
pub enum IssueState {
    #[allow(dead_code)]
    Open,
    #[allow(dead_code)]
    Closed,
    Merged,
    All,
}
```

This makes it still lint a warning if `IssueState` is never used at all, but ignore any warnings about `Open` and `Closed` specifically.

File: generate-release/src/github_client.rs
Comment (BD103) [inline generate-release/src/github_client.rs:409]:
```suggestion
    fn as_str(&self) -> &'static str {
```

It may be worth marking this explicitly `'static` so consumers know that the results are constant. On the other hand, it doesn't matter much because the lifetime is elided. Your choice 🤷‍♂️

File: generate-release/src/main.rs
Comment (BD103) [inline generate-release/src/main.rs:61]:
```suggestion
    /// While duplicate issues will not be opened, be sure to use the `--dry-run` flag
```

Adding the double-dash makes it more clear that this flag is supposed to be used in the CLI, not somewhere else. This is a stylistic choice, though, and I won't block on it. :)

File: generate-release/src/release_notes.rs
Comment (BD103) [inline generate-release/src/release_notes.rs:9]:
```suggestion
    io::Write as _,
```

We never mention `IoWrite` explicitly, so we don't need to name it in the imports.

**Review by BD103 — APPROVED** [review]
Looks good! I have a few nits, but they're inconsequential stylistic opinions that can be skipped :)

**Comment by alice-i-cecile** [comment]
I'm going to swap to leaving a comment in the merged PRs.

File: generate-release/src/release_notes.rs
Comment (alice-i-cecile) [reply generate-release/src/release_notes.rs:9]:
```suggestion
    io::Write,
```

File: generate-release/src/release_notes.rs
Comment (MrGVSV) [inline generate-release/src/release_notes.rs:202]: Nit: This wording is a bit strange to me since it almost sounds like its intended target is someone who both made and reviewed the PR (that might just be my take though lol). I also feel like the comment might come across as a demand rather than a polite request/suggestion.

Maybe we could do something like this?
```suggestion
        let comment = format!("Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to {issue_url} if you'd like to help out.",);
```

File: generate-release/src/release_notes.rs
Comment (MrGVSV) [inline generate-release/src/release_notes.rs:188]: Nit: Could we include the comment output here as well?

Comment (alice-i-cecile) [reply generate-release/src/release_notes.rs:188]: Not really unfortunately. The critical information (which issue to link to) isn't available until we actually create the issue and get the response, which we can then parse for the link.

Comment (MrGVSV) [reply generate-release/src/release_notes.rs:188]: Makes sense! It can always be added in a future PR (i.e. printing out the comment with a dummy link or something).

**Review by MrGVSV — APPROVED** [review]

File: generate-release/src/github_client.rs
Comment (kristoff3r) [inline generate-release/src/github_client.rs:245]:
```suggestion
    /// Request issues and PRs by the page returned by the Github API
```

**Review by kristoff3r — APPROVED** [review]

File: generate-release/src/main.rs
Comment (djeedai) [inline generate-release/src/main.rs:69]: nit: this will still overwrite any previously generated release notes, _e.g._ if you run it twice in a row. So it's not really a dry run (which normally means "no mutation").

File: generate-release/src/release_notes.rs
Comment (djeedai) [inline generate-release/src/release_notes.rs:205]: Can we avoid the `unwrap()` and instead log an error to whoever is running that command? Management scripts crashing is a plague of software maintenance 😅

**Review by djeedai — APPROVED** [review]
Only minor comments, looks good to me

File: generate-release/src/github_client.rs
Comment (alice-i-cecile) [reply generate-release/src/github_client.rs:188]: Agreed, but I don't want to touch this further in this PR.

File: generate-release/src/github_client.rs
Comment (alice-i-cecile) [reply generate-release/src/github_client.rs:453]: Agreed!

File: generate-release/src/release_notes.rs
Comment (alice-i-cecile) [reply generate-release/src/release_notes.rs:192]: Agreed.

File: generate-release/src/release_notes.rs
Comment (alice-i-cecile) [reply generate-release/src/release_notes.rs:205]: Can do!

File: generate-release/src/github_client.rs
Comment (BD103) [inline generate-release/src/github_client.rs:127]:
```suggestion
    /// Submits a GET request to `bevyengine/{repo}`
```

There are multiple HTTP requests types, such as GET, POST, DELETE, etc. This makes it clear which method is used.

File: generate-release/src/main.rs
Comment (BD103) [reply generate-release/src/main.rs:69]: `--no-issues` is another option, though I don't have a strong preference.

**Review by BD103 — APPROVED** [review]
Awesome! I have a few small nits, but overall 👍.

**Comment by alice-i-cecile** [comment]
@djeedai I've done one better than constants: I've swapped to an `BevyRepo` enum and changed the signatures to make it way harder to accidentally mess up.

File: generate-release/src/main.rs
Comment (BD103) [inline generate-release/src/main.rs:62]:
```suggestion
    /// While duplicate issues will not be opened, be sure to use the `--local` flag
```

**Comment by djeedai** [comment]
> @djeedai I've done one better than constants: I've swapped to an `BevyRepo` enum and changed the signatures to make it way harder to accidentally mess up.

C++ got me again! This was obviously the better solution! Thanks

**Review by djeedai — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

**Pull request merged**
